lib/devise/multi_email/models/authenticatable.rb

```diff
@@ -29,41 +29,73 @@ module AuthenticatableExtensions
         extend ActiveSupport::Concern
 
         included do
+          multi_email_association.configure_autosave!{ include AuthenticatableAutosaveExtensions }
           multi_email_association.include_module(EmailAuthenticatable)
         end
 
         delegate :skip_confirmation!, to: Devise::MultiEmail.primary_email_method_name, allow_nil: false
 
         # Gets the primary email address of the user.
         def email
-          multi_email.primary_email.try(:email)
+          multi_email.primary_email_record.try(:email)
         end
 
         # Sets the default email address of the user.
         def email=(new_email)
-          multi_email.change_primary_email_to(new_email)
+          multi_email.change_primary_email_to(new_email, allow_unconfirmed: true)
+        end
+      end
+
+      module AuthenticatableAutosaveExtensions
+        extend ActiveSupport::Concern
+
+        included do
+          primary_column = connection.quote_column_name(:primary)
+          id_column      = connection.quote_column_name(:id)
+
+          # Toggle `primary` value for all emails if `autosave` is not on
+          after_save do
+            if multi_email.primary_email_record
```

Comment:
We could loop through all email records and check if the `primary` attribute value was changed on any of them. I just pushed an update that calls `email.save` if the changes include `primary`. It's simplest this way because otherwise, the "dirty" state does not update after SQL updates are made, and calling `reload` on the email model would trigger another query. This solution should only run 2 update queries - to update the new primary and old primary records.